### ISSUE TITLE
Lookup problems with gem which integrates zeitwerk

### ISSUE THREAD

**Issue by fschwahn** [fxn/zeitwerk#300]
I'm using https://github.com/mbleigh/acts-as-taggable-on in a rails project (version 7.1.4 at the moment). I also use the pundit gem for authorization, and it expects that policies are defined as follows:
If I have a `Post`-model, it expects a `PostPolicy`-class to be defined which contains the authorization logic (see https://github.com/varvet/pundit?tab=readme-ov-file#policies for more).

`acts-as-taggable-on` has just released a new version, in which it adopts zeitwerk for autoloading (in this PR I believe: https://github.com/mbleigh/acts-as-taggable-on/pull/1132).

So, in order to use `pundit` and `acts-as-taggable-on` together, I define a class named `ActsAsTaggableOn::TagPolicy` in my codebase. This works when the app initially boots, but as soon as code reloads, it can no longer find `ActsAsTaggableOn::TagPolicy`, raising a NameError (it's enough to open a console and issue a `reload!`). When I downgrade `acts-as-taggable-on` to before it switched to zeitwerk, the problem goes away.

Is this a fundamental limitation when a gem uses zeitwerk for autoloading itself, or is something off with how this was implemented in `acts-as-taggable-on`?

**Comment by fxn** [comment]
Usage of Zeitwerk in that gem seems correct at first sight.

Could you please add `Rails.autoloaders.log!` to `config/application.rb`, reproduce, and share the traces?

**Comment by fschwahn** [comment]
Is it possible that it is connected to this custom inflection rule?

https://github.com/seuros/acts-as-taggable-on/blob/31f29c991e0cf6ec178e618be64e2b92f2f0b02e/lib/acts-as-taggable-on.rb#L6-L8

I truncated the log to only include the parts which I believe are relevant, I hope I didn't leave anything important out.

```
➜  app git:(master) ✗ rails c
Zeitwerk@rails.main: autoload set for ApplicationPolicy, to be loaded from /app/app/policies/application_policy.rb
Zeitwerk@rails.main: autoload set for ActsAsTaggableOn::TagPolicy, to be loaded from /app/app/policies/acts_as_taggable_on/tag_policy.rb
Running via Spring preloader in process 4414
Loading development environment (Rails 7.1.3.4)
irb(main):001> ActsAsTaggableOn::TagPolicy
Zeitwerk@rails.main: constant ApplicationPolicy loaded from file /app/app/policies/application_policy.rb
Zeitwerk@rails.main: constant ActsAsTaggableOn::TagPolicy loaded from file /app/app/policies/acts_as_taggable_on/tag_policy.rb
=> ActsAsTaggableOn::TagPolicy
irb(main):002> reload!
Reloading...
Zeitwerk@rails.main: ApplicationPolicy unloaded
Zeitwerk@rails.main: ActsAsTaggableOn::TagPolicy unloaded
Zeitwerk@rails.main: autoload set for ApplicationPolicy, to be loaded from /app/app/policies/application_policy.rb
  ActiveRecord::SchemaMigration Load (1.2ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
  ActiveRecord::SchemaMigration Load (1.0ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
=> true
irb(main):003> ActsAsTaggableOn::TagPolicy
(irb):3:in `<main>': uninitialized constant ActsAsTaggableOn::TagPolicy (NameError)

ActsAsTaggableOn::TagPolicy
                ^^^^^^^^^^^
Did you mean?  StatsPolicy
               RatePolicy
irb(main):004> 
```

Notably, the line

> Zeitwerk@rails.main: autoload set for ActsAsTaggableOn::TagPolicy, to be loaded from /app/app/policies/acts_as_taggable_on/tag_policy.rb

only appears on the initial app load, but not when reloading.

**Comment by fxn** [comment]
In principle, the inflection rule looks good. The namespace `ActsAsTaggableOn` is not managed by the Rails loader.

Do the logs mention `ActsAsTaggableOn` saying that the namespace already exists? I am missing those traces?

**Comment by fschwahn** [comment]
There are a bunch of those messages, such as

> Zeitwerk@rails.once: the namespace Turbo already exists, descending into /Users/fabian/.rbenv/versions/3.3.4/lib/ruby/gems/3.3.0/gems/turbo-rails-1.1.1/app/channels/turbo

However, there is no such message for `ActsAsTaggableOn`

**Comment by fxn** [comment]
Cool, I could reproduce. Let me investigate.

**Comment by fxn** [comment]
It was a bug, an edge case that happened specifically when 3rd party code reopened the main namespace of a gem dependency.

I have released 2.6.18 with the fix. Can you give it a try?
